database_admin/migrations/41_partition_packages.up.sql

```diff
@@ -0,0 +1,62 @@
+ALTER TABLE system_package
+    DROP CONSTRAINT system_package_package_id_fkey;
+
+
+ALTER TABLE system_package
+    DROP CONSTRAINT system_package_rh_account_id_fkey;
+
+
+DO
+$$
+    DECLARE
+        row text;
+    BEGIN
+        FOR row IN (SELECT tablename from pg_tables t where t.tablename like 'system_package%')
+            LOOP
+                EXECUTE 'ALTER TABLE ' || row || ' rename to ' || concat('old_', row);
+            END LOOP;
+    END
+$$ LANGUAGE plpgsql;
+
+DO
+$$
+    DECLARE
+        row text;
+    BEGIN
+        FOR row IN (SELECT indexname from pg_indexes t where t.indexname like 'system_package%')
+            LOOP
+                EXECUTE 'ALTER INDEX ' || row || ' rename to ' || concat('old_', row);
+            END LOOP;
+    END
+$$ LANGUAGE plpgsql;
+
+
+CREATE TABLE IF NOT EXISTS system_package
+(
+    rh_account_id INT NOT NULL REFERENCES rh_account,
+    system_id     INT NOT NULL,
+    package_id    INT NOT NULL REFERENCES package,
+    -- Use null to represent up-to-date packages
+    update_data   JSONB DEFAULT NULL,
+    latest_evra   TEXT GENERATED ALWAYS AS ( ((update_data ->> -1)::jsonb ->> 'evra')::text) STORED,
+    PRIMARY KEY (rh_account_id, system_id, package_id) INCLUDE (latest_evra)
+) PARTITION BY HASH (rh_account_id);
+
+
+GRANT SELECT, INSERT, UPDATE, DELETE ON system_package TO evaluator;
+GRANT SELECT, UPDATE, DELETE ON system_package TO listener;
+GRANT SELECT, UPDATE, DELETE ON system_package TO manager;
+GRANT SELECT, UPDATE, DELETE ON system_package TO vmaas_sync;
+
+SELECT create_table_partitions('system_package', 80,
+                               $$WITH (fillfactor = '70', autovacuum_vacuum_scale_factor = '0.05')$$);
+
+GRANT SELECT ON ALL TABLES IN SCHEMA public TO evaluator;
+GRANT SELECT ON ALL TABLES IN SCHEMA public TO listener;
+GRANT SELECT ON ALL TABLES IN SCHEMA public TO manager;
+GRANT SELECT ON ALL TABLES IN SCHEMA public TO vmaas_sync;
+
+INSERT INTO system_package (SELECT rh_account_id, system_id, package_id, update_data FROM old_system_package)
+ON CONFLICT (rh_account_id, system_id,package_id) DO NOTHING;
```

Comment:
how long it will take to copy whole table? will our db be even able to do it at once?

Reply:
Well, it'll be a long time, but I think we can make it. I could split it up to copying from individual partitions.